vendors/nordic/boards/nrf52840-dk/ports/common_io/iot_gpio.c

```diff
@@ -0,0 +1,622 @@
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ * http://aws.amazon.com/freertos
+ * http://www.FreeRTOS.org
+ */
+
+/**
+ * @file iot_gpio.c
+ * @brief HAL GPIO implementation on NRF52840 Development Kit. Note, CommonIO GPIO pin-index-space
+ *        is separate from board-specific pin index space. User is expected to provide a 1-to-1 mapping
+ *        between these spaces.
+ *
+ */
+
+#include "FreeRTOS.h"
+
+/* Nordic Board includes. */
+#include "nrf_gpio.h"
+#include "nrf_drv_gpiote.h"
+
+/* Common IO includes */
+#include "iot_gpio.h"
+#include "iot_gpio_config.h"
+
+/* Logging Includes */
+#include "iot_logging_task.h"
+
+/* Note, this depends on logging task being created and running */
+#if ( IOT_GPIO_LOGGING_ENABLED == 1 )
+    #define IOT_GPIO_MODULE_NAME    "[CommonIO][GPIO]"
+    #define IOT_GPIO_LOGF( format, ... )    vLoggingPrintf( IOT_GPIO_MODULE_NAME format, __VA_ARGS__ )
+    #define IOT_GPIO_LOG( msg )             vLoggingPrintf( IOT_GPIO_MODULE_NAME msg )
+#else
+    #define IOT_GPIO_LOGF( format, ... )
+    #define IOT_GPIO_LOG( msg )
+#endif
+
+
+#define IOT_GPIO_CLOSED    ( ( uint8_t ) 0 )
+#define IOT_GPIO_OPENED    ( ( uint8_t ) 1 )
+
+#define DEFAULT_GPIO_DESCRIPTOR                    \
+    {                                              \
+        .lGpioNumber = -1,                         \
+        .xConfig =                                 \
+        {                                          \
+            .xDirection     = eGpioDirectionInput, \
+            .xOutMode       = eGpioPushPull,       \
+            .lDriveStrength = NRF_GPIO_PIN_S0S1,   \
+            .xPull          = eGpioPullNone,       \
+            .xInterruptMode = eGpioInterruptNone   \
+        },                                         \
+        .xUserCallback = NULL,                     \
+        .pvUserContext = NULL,                     \
+        .ucState = IOT_GPIO_CLOSED                 \
+    }
+
+typedef struct
+{
+    IotGpioDirection_t xDirection;
+    IotGpioOutputMode_t xOutMode;
+    int32_t lDriveStrength;
+    IotGpioPull_t xPull;
+    IotGpioInterrupt_t xInterruptMode;
+} IotGpioConfig_t;
+
+/**
+ * @brief   GPIO descriptor type defined in the source file.
+ */
+typedef struct IotGpioDescriptor
+{
+    int32_t lGpioNumber;
+    IotGpioConfig_t xConfig;
+    IotGpioCallback_t xUserCallback;
+    void * pvUserContext;
+    uint8_t ucState;
```

Comment:
While enum is certainly better for readability, it wouldn't work well for masking states together like `ucState |= IOT_GPIO_BUSY` -- in case we add that in the future. Enums would also use word's worth of memory, versus a single byte